keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,432 @@
+
+### Goals
+
+TODO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+Adopt json-schema as the default format for `params.yaml`.
+
+- Json-Schema allows all of our requirements to be fulfilled
+- It is an existing standard
+- It can be expressed as yaml to fit into the Kubernetes ecosystem
+- It can be extended with custom keywords to define our custom attributes like `trigger` and `immutable`
+- There are existing go libraries 
+
+### Overview
+
+ - KUDO operator package format
+   The package format has to be adjusted. We need a new version, and KUDO should be able to read both old and new formats.
+- CRDS
+- Instance Updates `kudo update ... -p KEY=value`
+- Values.yaml `kudo update ... -P values.yaml` 
+- Add Parameter Wizard `kudo package add parameter`
+- Parameter Listing`kudo package list parameters`
+  We need a new output format here - maybe a tree-table?
+- Additional attributes in json-schema
+  We want to use json-schema to describe the parameter structure, but we need additional information that is not described json-schema.
+   
+### Package Format
+
+TODO: Describe new package format (json-schema)
+
+### CRDs
+
+Normally, per Kubernetes conventions, CRDs would evolve over released versions without a breaking change from one version to the next, allowing clients to gradually convert to new CRD versions.
+
+With this change, this will not be possible, as json-schema supports a lot of configuration which will not be mappable from a flat list.
+
+There will be a hard break from the current from `v1beta1` to the new `v1beta2`. We will provide a migration path for existing installed CRDs, but the clients accessing these CRDs will need to be updated.
+
+#### Conversion
+
+To migrate the existing CRDs, we will implement a Conversion Webhook that only allows `v1beta1` to `v1beta2` conversion, not the other way around. When KUDO is upgraded, it will:
+
+- Run Pre-Migration
+    - Fetch all existing CRDs, run the conversion and check for any errors
+- Deploy the new CRD version
+    - `v1beta1` will be still in the list, but with `storage=false, served=false`
+    - `v1beta2` will be marked as `storage=true, served=true`
+- Deploy the new manager with the conversion webhook
+- Run the migration
+    - Fetch all CRDs and issue an update - This will trigger a conversion and save the CRD with the new version
+    - Update the CRD status.storedVersions field
+- With the following KUDO release, we can remove `v1beta1` from the CRD list.
+
+Reference: [CRD Versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/)
+
+#### Operator
+
+The Operator CRD will not change, but we should keep the version in sync with the other CRDs.
+
+#### OperatorVersion
+
+In `v1beta1` the OV will contain a json-structure. This structure is mapped in the CRD to an `interface{}` type and will be read and interpreted by a go library.
```

Comment:
IIRC, `interface{}` isn't supported by `controller-gen`. I ran into this problem when adding optional types. The user facing `params.yaml` can use `interface{}` though, but this has to be converted to a string for the CRDs. An implementation detail to keep in mind 😄

Reply:
Yep, I just stumbled upon this. But you can use `unstructured.Unstructured` as a type, although that's a `map[string]interface{}`, but that should be ok.

Reply:
This needs more work. `unstructured.Unstructured` does not work either, as that expects the whole GVK in the structure for some reason. Maybe we have to work with `[]byte`, maybe we can add a typed structure for JSON schema.

Reply:
This is resolved now, we have to use `extensionapi.JSON`